### DISCUSSION TITLE
How to use APSW with other SQLite-using C extensions

### DISCUSSION THREAD

**Discussion by Nikratio** [rogerbinns/apsw#424] [Q&A] [answered]
Hi,

If I am using APSW together with another C extension that uses SQLite (say, an extension that defines a custom VFS), do I need to take any special care at build time?

For example, do I need to make sure that both are linked dynamically against the same SQLite library? Or is there some way to make things work even if APSW has SQLite compiled in statically?

**Comment by rogerbinns** [comment, marked as answer]
You don't need to do anything special and it should all work.  The only gotcha is that extension loading is off by default, and you need to enable it - eg with [enableloadextension](https://rogerbinns.github.io/apsw/connection.html#apsw.Connection.enableloadextension).  It is also the case that the default compilation of SQLite has loadable extension functionality compiled out, but virtually everyone has the functionality compiled in.

Loadable extensions are not linked against SQLite.  Instead the initialisation method of the extension is passed a list of function pointers.  That guarantees the correct functions are used no matter how and when the extension was compiled, and even if SQLite is compiled in statically.  You can see that [here](https://github.com/sqlite/sqlite/blob/master/src/sqlite3ext.h).

**Reply by Nikratio** [reply]
Thanks, that is encouraging!

What I still wonder is: how do I get APSW (or rather, the sqlite3 code that it uses) to call the initialisation method of my extension ?

**Reply by rogerbinns** [reply]
Apologies for not seeing this.  When SQLite loads an extension, it calls an init function `int
sqlite3_extension_init(sqlite3 *db, char **pzErrMsg, const sqlite3_api_routines *pApi)` that sets up the api pointers, lets you return an error message, and do whatever per connection work you need to do.  If that name doesn't exist, then it tries a similar name but using the basename of the extension file.

When you call [Connection.load_extension](https://rogerbinns.github.io/apsw/connection.html#apsw.Connection.load_extension) you can provide a second parameter with an alternate init function name to use.

This is somewhat covered in the [doc](https://sqlite.org/loadext.html) with section 5 covering how to persistently load your extension, and section 6 covering statically compiling alongside SQLite.

The most convenient thing to do if compiling your own SQLite is to append your extension to the amalgamation and `#define SQLITE_EXTRA_AUTOEXT` to be the name of your init function that only takes the `db` as a parameter.  It isn't documented and you'll need to look in the amalgamation source for more info.

**Comment by rogerbinns** [comment]
You can call [Connection.loadextension](https://rogerbinns.github.io/apsw/connection.html#apsw.Connection.loadextension) or [SQL load_extension()](https://www.sqlite.org/lang_corefunc.html#load_extension) function.

The standard [SQLite doc](https://www.sqlite.org/loadext.html) applies.  [This](https://github.com/rogerbinns/apsw/blob/master/src/testextension.c) is the test extension used when verifying the functionality with APSW.    [This](https://github.com/rogerbinns/apsw/blob/master/setup.py#L141) is the code used to build the test extension which ensures it uses the same compiler and settings as what was used to build Python.
